src/BaseDepository.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace Friendica;
+
+use Exception;
+use Friendica\Capabilities\ICanCreateFromTableRow;
+use Friendica\Database\Database;
+use Friendica\Database\DBA;
+use Friendica\Network\HTTPException\NotFoundException;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Depositories are meant to store and retrieve Entities from the database.
+ *
+ * The reason why there are methods prefixed with an underscore is because PHP doesn't support generic polymorphism
+ * which means we can't direcly overload base methods and make parameters more strict (from a parent class to a child
+ * class for example)
+ *
+ * Similarly, we can't make an overloaded method return type more strict until we only support PHP version 7.4 but this
+ * is less pressing.
+ */
+abstract class BaseDepository
```

Comment:
Slightly off topic, but: I do hope that at the hackathon you or @MrPetovan can explain in easy words what this whole stuff means. I see you using names that I keep forgetting faster than I can read a sentence. I guess it is as if I would explain the difference between a closed loop cycle and an open loop cycle (with the differences of being oxygen rich or fuel rich) in rocket engines when all that you know about rocket engines is that on the one side fuel goes in and on the side the exhaust goes out.

Reply:
anemic domain model (Depository): entities are some kind of type-safe readonly arrays, the Depository knows how to save/select them, the "Model" classes do the whole logic stuff

DDD (Repository): entities are "big" objects which holds the whole logic inside them (encapsualted) the Repositories know how to save/select them. So all relevant logic is bound to the entities itself which should make them more robust against changes. There is no need for model-logic classes

Reply:
The Depository is more applicable to the current codebase as far as I understand @mrpetovan because we already have model classes for logic and arrays for data.. It's less invasive.. The DDD means that we have to re-think the whole way how and where we put logic

Reply:
Thank you for the link. My goal isn't to have an anemic domain model. I want the Entities to perform business logic using and changing their properties. However, I want to keep static calls out of Entities until they have been replaced by the corresponding Entities/Collections/Factories/Depositories, so right now it's a little light by constraint, not by choice.

What I don't want is to be able to arbitrarily changing Entity properties. Each change should be associated with a specific use case. For example, for notifications from the moment one is created, no field but "seen" needs to change from `false` to `true`, so there's exactly one Entity method which task is to do just that.

This hardens the code against unexpected behaviors because we have consistent objects from their creation that can go through a very limited set of transformations.